primitives/src/block.rs

```diff
@@ -201,12 +201,40 @@ impl Header {
         engine.input(self.merkle_root.as_byte_array());
         engine.input(&self.time.to_u32().to_le_bytes());
         engine.input(&self.bits.to_consensus().to_le_bytes());
-        engine.input(&self.nonce.to_le_bytes());
+        engine.input(&self.nonce.to_u32().to_le_bytes());
 
         BlockHash::from_byte_array(sha256d::Hash::from_engine(engine).to_byte_array())
     }
 }
 
+impl fmt::Display for Header {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if f.alternate() {
+            write!(
+                f,
+                "{}:{}:{}:{}:{}:{}",
+                self.version,
+                self.prev_blockhash,
+                self.merkle_root,
+                self.time,
+                self.bits,
+                self.nonce
+            )
+        } else {
+            write!(
+                f,
+                "{}{}{}{}{}{}",
+                self.version,
+                self.prev_blockhash,
+                self.merkle_root,
+                self.time,
+                self.bits,
+                self.nonce
+            )
```

Comment:
We can easily do better here. We know the exact length so we could output the components into `ArrayString<160>` (or `ArrayString<165>` above) and then pass it to formatter to take care of formatting options.

Reply:
Cool. I think @apoelstra wants a custom implementation of the `Encodable` logic inside `Dislpay` instead of calling through to `Display` on all the fields. Right Andrew?

Reply:
Yeah, that's how I'd approach this. But what I "want" is whatever's easiest to implement and review :).

Reply:
No sweat.